toolchain/check/deduce.cpp

```diff
@@ -39,11 +39,31 @@ class DeductionWorklist {
         {.param = param, .arg = arg, .needs_substitution = needs_substitution});
   }
 
+  // Adds a single (param, arg) type deduction.
+  auto Add(SemIR::TypeId param, SemIR::TypeId arg, bool needs_substitution)
+      -> void {
+    Add(context_.types().GetInstId(param), context_.types().GetInstId(arg),
+        needs_substitution);
+  }
+
+  // Adds a single (param, arg) deduction of a specific.
+  auto Add(SemIR::SpecificId param, SemIR::SpecificId arg,
+           bool needs_substitution) -> void {
+    auto& param_specific = context_.specifics().Get(param);
+    auto& arg_specific = context_.specifics().Get(arg);
+    if (param_specific.generic_id != arg_specific.generic_id) {
+      // TODO: Decide whether to error on this or just treat the specific as
+      // non-deduced. For now we treat it as non-deduced.
+      return;
+    }
+    AddAll(param_specific.args_id, arg_specific.args_id, needs_substitution);
+  }
+
   // Adds a list of (param, arg) deductions. These are added in reverse order so
   // they are popped in forward order.
-  auto AddAll(llvm::ArrayRef<SemIR::InstId> params,
-              llvm::ArrayRef<SemIR::InstId> args, bool needs_substitution)
-      -> void {
+  template <typename InnerId>
+  auto AddAll(llvm::ArrayRef<InnerId> params, llvm::ArrayRef<InnerId> args,
```

Comment:
```suggestion
  template <typename ElementId>
  auto AddAll(llvm::ArrayRef<ElementId> params, llvm::ArrayRef<ElementId> args,
```

Minor, but maybe ElementId instead?